README.md
# bedrock-notifications

[![Build Status](http://ci.digitalbazaar.com/buildStatus/icon?job=bedrock-notifications)](http://ci.digitalbazaar.com/job/bedrock-notifications)

User notifications system for bedrock.

## Requirements

npm v3+


CHANGELOG.md
# bedrock-notifications ChangeLog

## 0.1.4 - 2016-11-08

### Changed
- Remove bedrock-messages-push dependency.

## 0.1.3 - 2016-09-22

### Changed
- Restructure test framework for CI.

## 0.1.2 - 2016-08-09

### Changed
- Namespace Protractor pages.

## 0.1.1 - 2016-07-07

### Changed
- Add tests.
- Properly encode URL on post.

## 0.1.0 - 2016-04-29

- See git history for changes.


test/test.js
/*
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var bedrock = require('bedrock');
require('bedrock-notifications');

require('bedrock-test');
bedrock.start();


test/config.js
var path = require('path');

module.exports = function(bedrock) {
  var config = bedrock.config;
  if(!config.protractor) {
    return;
  }
  var protractor = config.protractor.config;
  var prepare =
    path.join(__dirname, 'protractor', 'prepare.js');
  protractor.params.config.onPrepare.push(prepare);
};


bedrock/test.js
/*
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
var bedrock = require('bedrock');
require('../lib/notifications.js');

bedrock.start();


lib/index.js
/*
 * Bedrock notifications module.
 *
 * This module exposes an API for sending, recieving
 * and querying a database of notifications.
 *
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
 /* jshint node: true */

'use strict';

var async = require('async');
var bedrock = require('bedrock');
var BedrockError = bedrock.util.BedrockError;
var brPassport = require('bedrock-passport');
var brPermission = require('bedrock-permission');
var config = bedrock.config;
var database = require('bedrock-mongodb');
var ensureAuthenticated = brPassport.ensureAuthenticated;
var rest = require('bedrock-rest');
var scheduler = require('bedrock-jobs');

require('bedrock-messages');
require('bedrock-views');

var userSettings = null;
var defaultUserSettings = {};
var registeredModules = [];

require('./config');

var PERMISSIONS = config.permission.permissions;

// Tried pushing the job only in api.register(), but it seems
// like we can't do that (Job will never get run when pushed in register).
// TODO: See if we can fix that (messages-client has example),
// but not a big deal for now.
bedrock.config.scheduler.jobs.push({
  id: 'notifications.jobs.immediate',
  type: 'notifications.jobs.immediate',
  schedule: 'R/PT1M',
  concurrency: config.notifications.numWorkers
});
bedrock.config.scheduler.jobs.push({
  id: 'notifications.jobs.daily',
  type: 'notifications.jobs.daily',
  schedule: 'R/P1D',
  concurrency: config.notifications.numWorkers
});

var logger = bedrock.loggers.get('app');

var api = {};
module.exports = api;

api.getId = function(actor, id, callback) {
  async.auto({
    checkPermissions: function(callback) {
      brPermission.checkPermission(
        actor, PERMISSIONS.MESSAGE_ACCESS, {resource: id}, callback);
    },
    find: ['checkPermissions', function(callback) {
      var q = {
        id: database.hash(id)
      };
      userSettings.findOne(q, function(err, results) {
        if(results !== null && results.value) {
          return callback(null, results.value);
        }
        callback(err, {});
      });
    }],
    update: ['find', function(callback, results) {
      var needsUpdate = false;
      var settings = results.find;
      for(var defaultSetting in defaultUserSettings) {
        if(!(defaultSetting in settings)) {
          // User has no setting for an installed module, add it and update
          settings[defaultSetting] = defaultUserSettings[defaultSetting];
          needsUpdate = true;
        }
      }
      if(!needsUpdate) {
        return callback(null, settings);
      }
      var q = {
        id: database.hash(id)
      };
      var u = {
        $set: {
          value: settings
        }
      };
      userSettings.update(q, u, {upsert: true}, function(err) {
        return callback(err, settings);
      });
    }]
  }, function(err, results) {
    callback(err, results.update);
  });
};

bedrock.events.on('bedrock.init', function(callback) {
  // Define immediate and daily jobs
  scheduler.define('notifications.jobs.immediate', function(job, callback) {
    async.each(registeredModules, function(module, callback) {
      var options = module.getOptions();
      var processOptions = {
        method: options.type,
        interval: 'immediate'
      };
      module.process(processOptions, function(err) {
        // If we error out on this job, any leftover messages will get sent on
        // a subsequent job.
        return callback(err);
      });
    }, function(err) {
      // TODO: Figure out if error handling here is worth it.
      callback();
    });
  });
  // Define daily job
  scheduler.define('notifications.jobs.daily', function(job, callback) {
    // TODO: Haven't actually tested that this gets fired once daily,
    // how can we do that?
    async.each(registeredModules, function(module, callback) {
      var options = module.getOptions();
      var processOptions = {
        method: options.type,
        interval: 'daily'
      };
      module.process(processOptions, function(err) {
        // If we error out on this job, any leftover messages will get sent on
        // a subsequent job.
        return callback(err);
      });
    }, function(err) {
      // TODO: Figure out if error handling here is worth it.
      callback();
    });
  });

  logger.debug('Creating notification settings collection.');
  async.auto({
    openCollections: function(callback) {
      database.openCollections(['notificationPushUserSettings'
      ], function(err) {
        if(!err) {
          userSettings = database.collections.notificationPushUserSettings;
        }
        callback(err);
      });
    },
    createIndexes: ['openCollections', function(callback) {
      database.createIndexes([{
        collection: 'notificationPushUserSettings',
        fields: {id: 1},
        options: {unique: true, background: false}
      }], callback);
    }]
  }, function(err) {
    callback(err);
  });
});

// add routes
bedrock.events.on('bedrock-express.configure.routes', function(app) {
  // FIXME (collier): what is the permissions model for this?
  app.get(
    config.notifications.endpoints.settings + '/:id', rest.when.prefers.ld,
    ensureAuthenticated, function(req, res, next) {
      api.getId(req.user.identity, req.params.id, function(err, results) {
        if(err) {
          return next(err);
        }
        res.json(results);
      });
    });

  app.post(
    config.notifications.endpoints.settings + '/:id', rest.when.prefers.ld,
    ensureAuthenticated, function(req, res, next) {
      if(req.body.id !== req.params.id) {
        return next(new BedrockError(
          'User ID mismatch.', 'UserIdMismatch',
          {httpStatusCode: 409, 'public': true}));
      }
      updateSettings(req.user.identity, req.body, function(err) {
        if(err) {
          return next(err);
        }
        res.sendStatus(204);
      });
    });
});

// Lower level modules register themselves with this module to enable their
// notification processing.
api.register = function(module) {
  var options = module.getOptions();
  config.notifications.registeredServices.push(options);

  registeredModules.push(module);

  // Populate default user settings for module
  if(options.defaultUserSettings) {
    defaultUserSettings[options.type] = options.defaultUserSettings;
  }
};

// Used in unit testing, but we can probably take advantage of this
// functionality somewhere else in the future.
api._unregister = function(module) {
  var index = registeredModules.indexOf(module);
  if(index > -1) {
    console.log("Removing module from notifications");
    registeredModules.splice(index, 1);
  }
};

// exposed for testing
api._updateSettings = function(actor, options, callback) {
  updateSettings(actor, options, callback);
};

// FIXME (collier): does this module need its own set of permssions?
function updateSettings(actor, options, callback) {
  async.auto({
    checkPermissions: function(callback) {
      brPermission.checkPermission(
        actor, PERMISSIONS.MESSAGE_ACCESS, {resource: options.id}, callback);
    },
    updateSettings: ['checkPermissions', function(callback) {
      var q = {
        id: database.hash(options.id)
      };

      var set = {
        id: database.hash(options.id)
      };
      for(var type in options) {
        if(type === 'id') {
          continue;
        }
        set['value.' + type] = options[type];
      }
      var u = {
        $set: set
      };
      userSettings.update(q, u, {upsert: true}, callback);
    }]
  }, function(err, results) {
    callback(err, results.updateSettings);
  });
}


components/notification-directive.js
/*!
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
define([], function() {

'use strict';

/* @ngInject */
function factory(brNotificationService, config) {
  return {
    restrict: 'E',
    scope: {
      userId: '=brUserId'
    },
    templateUrl: requirejs.toUrl(
      'bedrock-notifications/components/notification-directive.html'),
    link: Link
  };

  function Link(scope) {
    var model = scope.model = {};
    model.loading = true;
    model.registeredOptions =
      config.data['bedrock-notifications'].registeredServices;
    var defaultSettings = {};
    for(var key in model.registeredOptions) {
      var option = model.registeredOptions[key];
      defaultSettings[option.type] = {
        enable: false,
        interval: option.intervals[0].value
      };
    }
    model.settings = defaultSettings;
    var storedSettings = {};

    scope.$watch('userId', function() {
      init();
    });

    function init() {
      brNotificationService.get(scope.userId)
        .then(function(result) {
          storedSettings = result.data;
          // required so that angular extend does not link model.storedSettings
          // with model.settings
          var stored = angular.copy(storedSettings);
          var defaults = angular.copy(defaultSettings);
          angular.extend(model.settings, defaults, stored);
          model.loading = false;
          scope.$apply();
        });
    }

    model.update = function() {
      model.loading = true;
      var options = {id: scope.userId};
      for(var type in model.settings) {
        if(type === 'id') {
          continue;
        }
        var value = model.settings[type];
        options[type] = value;
      }
      brNotificationService.update(options)
        .then(function(result) {
          model.loading = false;
          storedSettings = angular.copy(model.settings);
          scope.$apply();
        });
    };

    model.cancel = function() {
      // stored settings might be an empty object
      var stored = angular.copy(storedSettings);
      var def = angular.copy(defaultSettings);
      angular.extend(model.settings, def, stored);
    };
  }
}

return {brPushNotification: factory};

});


schemas/notification.bedrock-notifications.js
/*
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
// var validation = require('bedrock-validation');

// Do we use this file anywhere?
var schemas = require('bedrock-validation').schemas;

var schema = {
  type: 'object',
  title: 'Event',
  properties: {
    '@context': {type: 'string', required: true},
    date: schemas.w3cDateTime({required: true}),
    recipient: {type: 'string', required: true},
    sender: {type: 'string', required: true},
    subject: {type: 'string', required: false},
    type: {type: 'string', required: true},
    content: {
      required: true,
      properties: {
        body: {type: 'string', required: false},
        holder: {type: 'string', required: true},
        link: {type: 'string', required: true}
      }
    }
  }
};

module.exports = function() {
  return schema;
};


lib/notifications.js
/*
 * Bedrock notifications module.
 *
 * This module exposes an API for sending, recieving
 * and querying a database of notifications.
 *
 * Copyright (c) 2015-2016 Digital Bazaar, Inc. All rights reserved.
 */
 /* jshint node: true */

'use strict';

var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var BedrockError = bedrock.util.BedrockError;
var brPassport = require('bedrock-passport');
var brPermission = require('bedrock-permission');
var ensureAuthenticated = brPassport.ensureAuthenticated;
var config = bedrock.config;
var rest = require('bedrock-rest');
var uuid = require('node-uuid').v4;
var database = require('bedrock-mongodb');
var scheduler = require('bedrock-jobs');

var brMessages = require('bedrock-messages');

var userSettings = null;
var defaultUserSettings = {};
var registeredModules = [];

require('./config');

var PERMISSIONS = config.permission.permissions;

// Tried pushing the job only in api.register(), but it seems
// like we can't do that (Job will never get run when pushed in register).
// TODO: See if we can fix that (messages-client has example),
// but not a big deal for now.
bedrock.config.scheduler.jobs.push({
  id: 'notifications.jobs.immediate',
  type: 'notifications.jobs.immediate',
  schedule: 'R/PT1M',
  concurrency: config.notifications.numWorkers
});
bedrock.config.scheduler.jobs.push({
  id: 'notifications.jobs.daily',
  type: 'notifications.jobs.daily',
  schedule: 'R/P1D',
  concurrency: config.notifications.numWorkers
});

// configure for tests
bedrock.events.on('bedrock.test.configure', function() {
  require('./test.config');
});

var logger = bedrock.loggers.get('app');

var debugTesting = false;

var api = {};
module.exports = api;

api.getId = function(actor, id, callback) {
  async.auto({
    checkPermissions: function(callback) {
      brPermission.checkPermission(
        actor, PERMISSIONS.MESSAGE_ACCESS, {resource: id}, callback);
    },
    find: ['checkPermissions', function(callback) {
      var q = {
        id: database.hash(id)
      };
      var o = _.extend({}, database.writeOptions, {new: true, upsert: true});
      userSettings.findOne(q, function(err, results) {
        if(results !== null && results.value) {
          return callback(null, results.value);
        }
        callback(err, {});
      });
    }],
    update: ['find', function(callback, results) {
      var needsUpdate = false;
      var settings = results.find;
      for(var defaultSetting in defaultUserSettings) {
        if(!(defaultSetting in settings)) {
          // User has no setting for an installed module, add it and update
          settings[defaultSetting] = defaultUserSettings[defaultSetting];
          needsUpdate = true;
        }
      }
      if(!needsUpdate) {
        return callback(null, settings);
      }
      var q = {
        id: database.hash(id)
      };
      var u = {
        $set: {
          value: settings
        }
      };
      userSettings.update(q, u, {upsert: true}, function(err, results) {
        return callback(err, settings);
      });
    }]
  }, function(err, results) {
    callback(err, results.update);
  });
};

bedrock.events.on('bedrock.init', function(callback) {
  // Define immediate and daily jobs
  scheduler.define('notifications.jobs.immediate', function(job, callback) {
    async.each(registeredModules, function(module, callback) {
      var options = module.getOptions();
      var processOptions = {
        method: options.type,
        interval: 'immediate'
      };
      module.process(processOptions, function(err, results) {
        // If we error out on this job, any leftover messages will get sent on
        // a subsequent job.
        return callback(err);
      });
    }, function(err) {
      // TODO: Figure out if error handling here is worth it.
      callback();
    });
  });
  // Define daily job
  scheduler.define('notifications.jobs.daily', function(job, callback) {
    // TODO: Haven't actually tested that this gets fired once daily,
    // how can we do that?
    async.each(registeredModules, function(module, callback) {
      var options = module.getOptions();
      var processOptions = {
        method: options.type,
        interval: 'daily'
      };
      module.process(processOptions, function(err, results) {
        // If we error out on this job, any leftover messages will get sent on
        // a subsequent job.
        return callback(err);
      });
    }, function(err) {
      // TODO: Figure out if error handling here is worth it.
      callback();
    });
  });

  logger.debug('Creating notification settings collection.');
  async.auto({
    openCollections: function(callback) {
      database.openCollections(['notificationPushUserSettings'
      ], function(err) {
        if(!err) {
          userSettings = database.collections.notificationPushUserSettings;
        }
        callback(err);
      });
    },
    createIndexes: ['openCollections', function(callback) {
      database.createIndexes([{
        collection: 'notificationPushUserSettings',
        fields: {id: 1},
        options: {unique: true, background: false}
      }], callback);
    }]
  }, function(err) {
    callback(err);
  });
});

bedrock.events.on('bedrock.ready', function(callback) {
  var brMessagesPush = require('bedrock-messages-push');

  bedrock.events.on('bedrock-messages.NewMessage', function(message) {
    if(debugTesting) {
      return;
    }
    brMessagesPush.queue.add(message, function(err) {
      if(err) {
        // TODO: Handle error state, can we recover from a failed add?
        console.log('Error pushing message onto notification queue', err);
      }
    });
  });
  callback();
});

// add routes
bedrock.events.on('bedrock-express.configure.routes', function(app) {
  // FIXME (collier): what is the permissions model for this?
  app.get(
    config.notifications.endpoints.settings + '/:id', rest.when.prefers.ld,
    ensureAuthenticated, function(req, res, next) {
      api.getId(req.user.identity, req.params.id, function(err, results) {
        if(err) {
          return next(err);
        }
        res.json(results);
      });
    });

  app.post(
    config.notifications.endpoints.settings + '/:id', rest.when.prefers.ld,
    ensureAuthenticated, function(req, res, next) {
      if(req.body.id !== req.params.id) {
        return next(new BedrockError(
          'User ID mismatch.', 'UserIdMismatch',
          {httpStatusCode: 409, 'public': true}));
      }
      updateSettings(req.user.identity, req.body, function(err, results) {
        if(err) {
          return next(err);
        }
        res.sendStatus(204);
      });
    });
});

// Lower level modules register themselves with this module to enable their
// notification processing.
api.register = function(module) {
  var options = module.getOptions();
  config.notifications.registeredServices.push(options);

  registeredModules.push(module);

  // Populate default user settings for module
  if(options.defaultUserSettings) {
    defaultUserSettings[options.type] = options.defaultUserSettings;
  }
};

// Used in unit testing, but we can probably take advantage of this functionality
// somewhere else in the future.
api._unregister = function(module) {
  var index = registeredModules.indexOf(module);
  if(index > -1) {
    console.log("Removing module from notifications");
    registeredModules.splice(index, 1);
  }
};

// exposed for testing
api._updateSettings = function(actor, options, callback) {
  updateSettings(actor, options, callback);
};

// FIXME (collier): does this module need its own set of permssions?
function updateSettings(actor, options, callback) {
  async.auto({
    checkPermissions: function(callback) {
      brPermission.checkPermission(
        actor, PERMISSIONS.MESSAGE_ACCESS, {resource: options.id}, callback);
    },
    updateSettings: ['checkPermissions', function(callback) {
      var q = {
        id: database.hash(options.id)
      };

      var set = {
        id: database.hash(options.id)
      };
      for(var type in options) {
        if(type === 'id') {
          continue;
        }
        set['value.' + type] = options[type];
      }
      var u = {
        $set: set
      };
      userSettings.update(q, u, {upsert: true}, callback);
    }]
  }, function(err, results) {
    callback(err, results.updateSettings);
  });
}

api._setDebugTesting = function() {
  debugTesting = true;
};
